cmd/controller-manager/main.go

```diff
@@ -119,44 +151,93 @@ func main() {
+			err = servicedns.StartController(controllerConfig, stopChan)
+			if err != nil {
+				glog.Fatalf("Error starting dns controller: %v", err)
+			}
+
+			err = dnsendpoint.StartServiceDNSEndpointController(controllerConfig, stopChan)
+			if err != nil {
+				glog.Fatalf("Error starting dns endpoint controller: %v", err)
+			}
 		}
-	}
 
-	if utilfeature.DefaultFeatureGate.Enabled(features.FederatedIngress) {
-		err = ingressdns.StartController(controllerConfig, stopChan)
-		if err != nil {
-			glog.Fatalf("Error starting ingress dns controller: %v", err)
+		if utilfeature.DefaultFeatureGate.Enabled(features.FederatedIngress) {
+			err = ingressdns.StartController(controllerConfig, stopChan)
+			if err != nil {
+				glog.Fatalf("Error starting ingress dns controller: %v", err)
+			}
+
+			err = dnsendpoint.StartIngressDNSEndpointController(controllerConfig, stopChan)
+			if err != nil {
+				glog.Fatalf("Error starting ingress dns endpoint controller: %v", err)
+			}
 		}
 
-		err = dnsendpoint.StartIngressDNSEndpointController(controllerConfig, stopChan)
-		if err != nil {
-			glog.Fatalf("Error starting ingress dns endpoint controller: %v", err)
+		if utilfeature.DefaultFeatureGate.Enabled(features.PushReconciler) {
+			federatedtypeconfig.StartController(controllerConfig, stopChan)
 		}
+
+		// Blockforever
+		select {}
 	}
 
-	if utilfeature.DefaultFeatureGate.Enabled(features.PushReconciler) {
-		federatedtypeconfig.StartController(controllerConfig, stopChan)
+	if !enableLeaderElection {
+		run(stopChan)
+		glog.Fatalf("finished without leader elect")
+	}
+	restclient.AddUserAgent(kubeConfig, "leader-election")
+	leaderElectionClient := clientset.NewForConfigOrDie(kubeConfig)
+	if err != nil {
+		glog.Fatalf("Error starting ingress dns endpoint controller: %v", err)
 	}
 
-	// Blockforever
-	select {}
+	// Prepare event clients.
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: leaderElectionClient.CoreV1().Events(fedNamespace)})
+	eventRecorder := broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "kar-controller"})
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		glog.Fatalf("unable to get hostname: %v", err)
+	}
+	// add a uniquifier so that two processes on the same host don't accidentally both become active
+	id := hostname + "_" + string(uuid.NewUUID())
+	rl, err := resourcelock.New(resourceLock,
+		fedNamespace,
+		"federation-controller",
+		leaderElectionClient.CoreV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      id,
+			EventRecorder: eventRecorder,
+		})
+	if err != nil {
+		glog.Fatalf("couldn't create resource lock: %v", err)
+	}
+
+	leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: leaseDuration,
+		RenewDeadline: renewDeadline,
+		RetryPeriod:   retryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
```

Comment:
The leader election callback would appear to ensure that a master is started, but how will a master be stopped if it loses the leadership?  It appears that a master once started would never be stopped since there is nothing that closes the stop channel.

Reply:
That suggests that the stop channel for the non-leader elect path be created in the `if !enableLeaderElection` so that it's clear that it won't be used outside of that block.

Reply:
@marun Have updated, now we can pass the stopChan through the ctx to the run func. I have checked in my environment, it works fine :) PTAL.